lib/srv/alpnproxy/conn_test.go

```diff
@@ -0,0 +1,361 @@
+					if bytes.Equal(dataWritten, aggregator) {
+						continue readLoop
+					}
+				}
+			}
+		}
+	})
+
+	t.Run("ConcurrentWrites", func(t *testing.T) {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+
+		// Create TCP connections to have asynchronous read/write.
+		w, r := makeTCPPingConn(t)
+		defer w.Close()
+		defer r.Close()
+
+		nWrites := 10
+		dataWritten := []byte("message")
+		writeChan := make(chan error)
+
+		// Start write routines.
+		for i := 0; i < nWrites/2; i++ {
+			go func() {
+				for writes := 0; writes < 2; writes++ {
+					err := w.WritePing()
+					if err != nil {
+						writeChan <- err
+						return
+					}
+
+					_, err = w.Write(dataWritten)
+					if err != nil {
+						writeChan <- err
+						return
+					}
+				}
+
+				writeChan <- nil
+			}()
+		}
+
+		// Expect all writes to succeed.
+		for i := 0; i < nWrites/2; i++ {
+			select {
+			case <-ctx.Done():
+				require.Fail(t, "timout write")
+			case err := <-writeChan:
+				require.NoError(t, err)
+			}
+		}
+
+		// Read all messages.
+		buf := make([]byte, len(dataWritten))
+		for i := 0; i < nWrites; i++ {
+			n, err := r.Read(buf)
+			require.NoError(t, err)
+			require.Equal(t, dataWritten, buf[:n])
+		}
+	})
+
+	t.Run("LargeContent", func(t *testing.T) {
+		t.Run("SingleRead", func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+			defer cancel()
+
+			r, w := makePingConn(t)
+			defer r.Close()
+			defer w.Close()
+
+			largeSlice := make([]byte, math.MaxUint32)
+			errChan := make(chan error, 2)
+
+			// Produce some data to it to match after reading.
+			written, err := rand.Read(largeSlice)
+			require.NoError(t, err)
+			require.Equal(t, math.MaxUint32, written)
+
+			// Write goroutine.
+			go func() {
+				_, err = w.Write(largeSlice)
+				errChan <- err
+			}()
+
+			// Read goroutine.
+			go func() {
+				buf := make([]byte, math.MaxUint32)
+				read, err := r.Read(buf)
+				if err != nil {
+					errChan <- fmt.Errorf("expected no read error but got %s", err)
+					return
+				}
+
+				if read != math.MaxUint32 {
+					errChan <- fmt.Errorf("expected to read %d but got %d", math.MaxUint32, read)
+					return
+				}
+
+				if !bytes.Equal(largeSlice, buf) {
+					errChan <- fmt.Errorf("expected to content to be the same")
+					return
+				}
+			}()
+
+			for i := 0; i < 1; i++ {
+				select {
+				case <-ctx.Done():
+					require.Fail(t, "exceeded time to read/write message")
+				case err := <-errChan:
+					require.NoError(t, err)
+				}
+			}
+		})
+
+		t.Run("LargerThanUint32", func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+
+			errChan := make(chan error)
+			_, w := makePingConn(t)
+			defer w.Close()
+
+			// Place into a goroutine to avoid freezing tests in case where the
+			// write succeeds.
+			go func() {
+				largeSlice := make([]byte, math.MaxUint32+1)
```

Comment:
is this 4g of memory? i would not dare to run this test.

Reply:
Yeah, these tests (with large inputs) were not intended to be executed with the unit testing suite. I considered:
1. removing them;
2. keeping them but only running when a flag/env var is provided;

What do you think?

Reply:
how about make it configurable so test ping conn can use a different default?

Reply:
"Regular sized" messages are well covered on other tests. These tests were done specifically to test these edge cases where the connections are exchanging this large-sized message. I'll add a `t.Skip` for now to give the PR unit tests a chance to complete, and we can decide if we want to keep them.

Reply:
I don't think we should allow that scenario in any case, especially in production. If you want to exchange more than 20MB probably, you should divide the message into chunks. If you run Teleport in a memory-restricted environment, then allocating a few GB of memory will probably kill the process because of OOM.
I suggest changing the max size of a single package to, let's say, 1MB, reduce the size here, and run this test always.